Fix s1 13188 pr

hakonfam opened the pull request:
https://github.com/SebastianBoe/fw-nrfconnect-zephyr/pull/1 is needed for this to work

rlubos committed:
zigbee: Make use of the L2_CUSTOM_IEEE802154 from Zephyr

Instead of relying on downstream patches in sdk-zephyr, make use of the
custom IEEE802154 L2 confguration to implment Zigbee L2.

Increase the expected ACK size, as the 15.4 shim layer will now inlcude
the FCS field in ACK for Zigbee.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

SebastianBoe committed:
manifest: Use an NCS fork of mbedtls instead of upstream mbedtls

Use an NCS fork of mbedtls instead of upstream mbedtls. No patches
yet, but they are expected soon to temporarily deal with the changes
in mbedtls 3.0.0.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
manifest: Add Zephyr module fff

Add the testing framework Zephyr module fff to the name-allowlist.

This could be required to run certain tests in the future.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
scripts: ncs_commands: add gigadevice to _BLOCKED_PROJECTS

Add the hal gigadevice to _BLOCKED_PROJECTS.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
manifest: Add Zephyr modules psa-arch-tests and tf-m-tests

Add the two TF-M Zephyr modules psa-arch-tests and tf-m-tests. These
used to be copied into the tfm module. But now exist as their own
modules.

These are required for integration testing of TF-M.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
manifest: Add Zephyr module zscilib

Add the scientific/math Zephyr module zscilib to the name-allowlist.

This could be a useful replacement for gsl / Lis / CMSIS-DSP.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
[nrf mergeup] manifest: zephyr: upmerge to jan. 5 SHA

Upmerge from 2.7.0 to the Zephyr main branch on 5. january.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
[nrf mergeup] manifest: mcuboot: upmerge jan. 5

Upmerge to 399720d1cabd26c4356445d351f263b31e942961, which was the
mcuboot revision used by Zephyr on January 5.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
manifest: Move Zephyr module trusted-firmware-m

tfm used to be a repository that had the contents of three
repositories copied into it. This tfm repository no longer exists. In
it's place are the three repositories that used to be copied into tfm.

trusted-firmware-m
psa-arch-tests
tf-m-tests

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
caf: leds: Replace deprecated DEVICE_POWER_MANAGEMENT

Use CONFIG_PM_DEVICE instead of the deprecated
CONFIG_DEVICE_POWER_MANAGEMENT.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
kconfig: Rename the deprecated LOG_MINIMAL to LOG_MODE_MINIMAL

LOG_MINIMAL has been deprecated so we replace all usage with
LOG_MODE_MINIMAL.

NB: LOG_MODE_MINIMAL is a choice and cannot be selected/implied. For
this use-case we select/imply LOG_DEFAULT_MINIMAL instead.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
power: reboot.h: Add power/reboot.h for compatibility

All code that does

needs to be ported to

This file maintains backwards compatibility until all users have
ported to the new include path.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
reboot: include <sys/reboot.h> instead of <power/reboot.h>

reboot.h was moved. Update the include paths to reflect this.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
upmerge: pm: Port to new pm_device_ API

There are some breaking changes to the power management API.

pm/device.h needs to be included. Presumably this came from device.h
before.

pm_device_state_set no longer exists. Nor does there exist a
one-to-one replacement for it. But pm_device_action_run is close.

Instead of requesting states, one requests actions, and hopes that the
desired state is achieved.

There is a new enum for describing the actions. This means that one
can no longer do.

function(target_state)
change_state(target_state)
endfunction()

Instead one must translate from the target state to some action, and
then do the action. Or drop support for requesting target states.

The device.h API does not support such a translation, so we would need
to translate ourselves.

In some cases we have been checking the current state, and then
comparing it to the target state before requesting the target
state. This is awkward with the new API.

In these cases we are now instead skipping the check of current state,
and just doing the action that should give us the desired state. If
the state was already set correctly we get a -EALREADY error code
which we suppress. This is nearly functionally the same as checking
the current state first.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
pm: pm_power_state_force: Port to new API

pm_power_state_force now requires a CPU index argument.

It is not clear how to determine what this should be, so we use
_current_cpu->id.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
[nrf mergeup] manifest: trusted-firmware-m: upmerge jan. 5

Upmerge from 1.3.99 to 1.5.0 (+ downstream patches).

The 1.3.99 revision contained 3 repo's, including
trusted-firmware-m. This new revision is the trusted-firmware-m
repo. So the repo looks completely different.

History has not been retained as it is in practice a new repository.

The sauce tag commits were re-applied to this new repository.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
shell: Ported breaking change in shell_init API

The shell_init function prototype has changed, instead of the third
argument being a bool, it is now a "struct
shell_backend_config_flags".

Port all usage to the new API.

Breaking change:
https://github.com/zephyrproject-rtos/zephyr/commit/
2b5723d4558619a9283683499c095e0b344d32d7

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
[nrf mergeup] manifest: sdk-mbedtls: Upmerge v3.0.0

Update sdk-mbedtls from 2.26.0 to 3.0.0.

3.0.0 is the same as zephyr's mbedtls except that zephyr has some
extra patches.

This updates ncs/mbedtls, as opposed to
ncs/modules/crypto/mbedtls.

Eventually, we will remove this so there are not so many copies of
mbedtls.

This update fixes some integration issues with nrf_security.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
drivers: timer: Port to new SYSTEM_CLOCK_DISABLE API

CONFIG_SYSTEM_CLOCK_DISABLE no longer exists.

It used to add a function to the kernel. In it's place is
CONFIG_SYSTEM_TIMER_HAS_DISABLE_SUPPORT.

This new option works slightly differently. It also adds the function
to the kernel, but instead of the user enabling it, it should be
selected by any provider of this kernel function.

The Kconfig option was changed in this commit:
https://github.com/zephyrproject-rtos/zephyr/commit/
78dc8ce33828b1d995d520751fe761ade6103155

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
manifest: Remove mcumgr which is now copied into Zephyr

Remove the mcumgr module which is now copied into Zephyr instead.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

gmarull committed:
doc: mcuboot: do not copy all .md files

We can't copy anymore all *.md files, since some of them reference now
documents outside of the docs/ folder. We have been lucky that none of
the required documents do this, however, this may happen in the future.
This shows that the MCUboot docset approach is not sustainable in the
long term, so a better way needs to be found.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: nrf: update to cbor2

The cbor dependency came from MCUboot, which now uses cbor2.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: nrf: gs_modifying: remove unnecessary information

The recent Zephyr docs describe exactly what is contained in the
paragraph above, other information has been removed.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

SebastianBoe committed:
manifest: nrfxlib: Ported to be compatible with new Zephyr revision

Bumped nrfxlib to a revision that is compatible with new Zephyr
revision.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

gmarull committed:
doc: nrf: libraries: caf: update PM API reference

The PM API reference is now pm_api.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: nrf: update MCUboot index references

The MCUboot index page is now mcuboot:index-ncs. Documentation was not
updated when the new index page was added. It shows up now because the
old index is not copied anymore.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: scripts: merge_search_indexes: update for Sphinx 4.x

The searchindex.js file format includes some incompatible changes with
respect to 3.x. This patch update the merge objects code to handle the
new format.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

kkasperczyk-no committed:
manifest: Updated Matter project revision to support mbedTLS 3.0.0

Changed sdk-connectedhomeip fork revision to support mbedTLS 3.0.0
version API in Matter.

Signed-off-by: Kamil Kasperczyk <kamil.kasperczyk@nordicsemi.no>

ankuns committed:
dts: bindings: remove generic-fem-two-ctrl-pins.yaml

The file generic-fem-two-ctrl-pins.yaml is provided by zephyr

Signed-off-by: Andrzej Kuros <andrzej.kuros@nordicsemi.no>

SebastianBoe committed:
zigbee: Fix printf formatting type issue

Fix printf formatting type issue. Perhaps delay_ms used to be signed.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe committed:
gpio: Don't use deprecated CONFIG_GPIO_NRF_P0

CONFIG_GPIO_NRF_P0 and friends no longer exists. Migrate all code off
of it.

See

https://github.com/zephyrproject-rtos/zephyr/commit/
cf3dabedaff535313e38e84e3a37b27ba3e815bb

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>
Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

rlubos committed:
zigbee: Get rid of unaligned access warnings

Rework the code to make it prepared for an unaliged access.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

gmarull committed:
doc: requirements: update

- Update theme version (to allow base theme >= 1.0)
- Remove pygments, already inherited from Zephyr

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

SebastianBoe committed:
samples: radio_test: Fix string formatting

These string formats were producing warnings.

config.mode is an enum and should therefore be formatted as %d.

channel_start and channel_end are unit8_t and should therefore not use
the l modifier.

It's not clear why lu was triggering a warning as the type is
uint32_t. Perhaps the correct modifier would have been llu, but u also
works for uint32_t.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

gmarull committed:
doc: nrf: templates: remove duplicate toctree entry

The toctree included 2 entries pointing to the same document. This seems
to raise a warning in new Sphinx versions.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

SebastianBoe committed:
power: Use CONFIG_PM instead of CONFIG_SYS_POWER_MANAGEMENT

CONFIG_SYS_POWER_MANAGEMENT is no longer supported. Use CONFIG_PM
instead.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

hakonfam committed:
cmake: extract function for generating cmake args for child images

Create separate function for generating a string of CMake arguments
that should be passed to a child image.

This improves readability of 'add_child_image' and facilitates future
improvements where other means of building the CMake arguments could
become relevant.

Ref: NCSDK-13188

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

hakonfam committed:
cmake: add function for generating variant build of an image

Also add function for generating the S1 variant specifically.

Ref: NCSDK-13188
Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

hakonfam committed:
cmake: use add_s1_variant() instead of s1.cmake

Leverage multi image mechanism for building s1 variant instead of
creating a new linker script and re-linking the base image.

Ref: NCSDK-13188
Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -23,125 +23,27 @@ if(IMAGE_NAME)
+#     CHILD_IMAGE            <name of child image>
+#     OUTPUT_CMAKE_ARGS      <output variable for storing cmake args>
+#   )
+function(get_child_image_cmake_args)
```
varargs should not be used when the number of arguments is not variable.

hakonfam replied:
ok, will fix

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -267,19 +170,146 @@ function(add_child_image_from_source)
   endforeach()
 
+  unset(${GET_ARGS_CONFIG_OUT} PARENT_SCOPE)
+  set(${GET_ARGS_CONFIG_OUT} "${child_image_cmake_args}" PARENT_SCOPE)
```
unset(a)
set(a foo)

is the same as

set(a foo)

no?

hakonfam replied:
I would think so, I'll remove it.

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -139,146 +398,20 @@ function(add_child_image_from_source)
+  set(${ACI_NAME}_IMAGE_CMAKE_ARGS "${image_cmake_args}" CACHE INTERNAL "")
 
   file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/${ACI_NAME})
   execute_process(
```
${EXTRA_MULTI_IMAGE_CMAKE_ARGS} # E.g. --trace-expand

which is below here looks scary.

How do we know people are not putting, e.g.

-DCONFIG_FOO=y

in there and accidentally creating a difference between S0 and S1 ?

hakonfam replied:
It is passed down since its a regular CLI arg.

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -185,7 +185,95 @@ function(get_child_image_cmake_args)
   set(${GET_ARGS_CONFIG_OUT} "${child_image_cmake_args}" PARENT_SCOPE)
+endfunction()
+
+function(add_s1_variant)
```
The args here do not appear to be variable.

hakonfam replied:
You can choose to not send `BASE_IMAGE`

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -185,7 +185,95 @@ function(get_child_image_cmake_args)
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+# SOURCE_DIR - The source dir of the base image.
+# EXTRA_CONFIG - The extra configuration fragment to pass to the variant image.
+function(add_image_variant)
```
Not varargs AFAICT

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -185,7 +185,95 @@ function(get_child_image_cmake_args)
+#
+# Required arguments are:
+# IMAGE_NAME - The name of the variant image
+# BASE_IMAGE - The name of the base image which the variant image is based on.
```
Not actually required?

hakonfam replied:
Correct, I will fix that

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -23,6 +23,259 @@ if(IMAGE_NAME)
+      get_property(var_help CACHE ${var_name} PROPERTY HELPSTRING)
+      if("${var_help}" STREQUAL "No help, variable specified on the command line.")
+        list(APPEND application_vars ${var_name})
+      endif()
```
Need a comment explaining this.

hakonfam replied:
Added comment in the top of the function

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -23,6 +23,259 @@ if(IMAGE_NAME)
+      "-DOVERLAY_CONFIG=${AIV_EXTRA_CONFIG}\\\\;"
+      variant_image_cmake_args
+      "${base_image_configuration}"
+      )
```
I see that config_files. Has the highest priority. Perhaps if we moved and renamed the parent's .config into the application's build dir with the name copy_of_parent.conf.

Then we wouldn't need to care about any of the other settings.

# Bring in extra configuration files dropped in by the user or anyone else;
# make sure they are set at the end so we can override any other setting
file(GLOB config_files ${APPLICATION_BINARY_DIR}/*.conf)
list(SORT config_files)
set(
  merge_config_files
  ${BOARD_DEFCONFIG}
  ${CONF_FILE_AS_LIST}
  ${shield_conf_files}
  ${OVERLAY_CONFIG_AS_LIST}
  ${EXTRA_KCONFIG_OPTIONS_FILE}
  ${config_files}
)

SebastianBoe replied:
In other words, instead of copying and modifying the configuration, we override it.

SebastianBoe replied:
Discussed offline, this will only reduce a small amount of the business logic, so it's not worth doing.

The new approach is still considered to be the lesser of the two evils.

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -185,7 +185,95 @@ function(get_child_image_cmake_args)
+  add_child_image(
+    NAME ${AIV_IMAGE_NAME}
+    SOURCE_DIR ${AIV_SOURCE_DIR}
+    )
```
There is a lot about Kconfig here.

What about DT?

Are there no variables controlling the DT sources ?

hakonfam replied:
Those are known when we build the image. All DT sources are picked from board name or user input, both of which is passed down to the variant image.

hakonfam replied:
Added handling of DTC_OVERLAY_FILE

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -23,6 +23,278 @@ if(IMAGE_NAME)
+
+    set(base_image_configuration "${${AIV_BASE_IMAGE}_IMAGE_CMAKE_ARGS}")
+  else()
+    # No base image is specified, this means that the base image is the top
```
Can't remember using the term base image, is it the same as parent image?

hakonfam replied:
It is introduced in the docs for these functions, it is the image which the variant is based upon.

SebastianBoe commented on cmake/multi_image.cmake:
```diff
@@ -23,6 +23,278 @@ if(IMAGE_NAME)
+    # inspecting their help text. However, some variables have special
+    # handling, resulting in a different help text. These cannot be found
+    # using the same mechanisms as the regular variables, and needs special
+    # handling.
```
Need to be more precise about this obscure behaviour with the help text.

Instead of

These can typically be identified by inspecting their help text.

we write

Command line arguments can generally be identified in the CMakeCache because they have the same help text generated by CMake. The text: "No help, variable specified on the command line."".

Alternatively we could have this comment closer to the code itself.

hakonfam replied:
Added this text closer to the code.

hakonfam committed:
cmake: support for setting full conf of child image through a variable

Also expose each child images' configuration through a cache variable.

This is useful for creating variant builds of an image.

Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

SebastianBoe commented on subsys/bootloader/image/CMakeLists.txt:
```diff
@@ -4,6 +4,10 @@
+if (CONFIG_IS_S1_IMAGE)
+  return()
+endif()
+
```
Need a comment explaining when subsys/bootloader/image/CMakeLists.txt is executed and why we don't need to run any of the below scripts when CONFIG_IS_S1_IMAGE.

SebastianBoe commented on cmake/s1.cmake:
```diff
@@ -1,153 +0,0 @@
-
-  configure_linker_script(
-    ${link_variant}linker_prebuilt.cmd
-    "-DLINK_INTO_${link_variant_name}"
```
Where is the code that modifies which area of flash S1 is linked into ?

hakonfam replied:
https://github.com/SebastianBoe/fw-nrfconnect-zephyr/pull/1/commits/dc49c027b351ad581441cbc840304d21757d2298#diff-54920465c2b794413e2858cbade114efdb0b29c59e83d7e188ba423887a2977aR35

hakonfam committed:
partition_manager: fix bug in bootlader/pm.yml

MCUboot child image kconfig options are not available when preprocesing
this file. Hence, checking if CONFIG_PARTITION_SIZE_MCUBOOT_PAD = 0
will always be true.

This places s1_image to an incorrect location.

Ref: NCSDK-13188
Signed-off-by: Håkon Øye Amundsen <haakon.amundsen@nordicsemi.no>

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -23,125 +23,15 @@ if(IMAGE_NAME)
+# Function for gathering all CMake arguments relevant to Kconfig and Devicetree
+# for a child image. The output can be passed directly to a CMake call.
+#
+function(get_child_image_cmake_args child_image output_cmake_args)
```
This seems to have become a purely internal function, but it looks like an external function from description and the description of child image configuration.

I find that very confusing so if we want an internal function, then we should look at how we structure the code and place the comment which is used `as-is` by sphinx for doc generation.

hakonfam replied:
I understand your concern, it looks like this function is supposed to be called from outside this file. The reason I pulled this out to its own function was to make the code inside `add_child_image` more readable. In particular this section:

```
  message("\n=== child image ${ACI_NAME} - ${ACI_DOMAIN}${inherited} begin ===")

  if (PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME})
    # A complete set of configurations has been specified. Use this instead of
    # finding the cmake args the normal way.
    set(image_cmake_args "${PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME}}")
  else()
    get_child_image_cmake_args(${ACI_NAME} image_cmake_args)
  endif()
```
Where it is easy for the user to see that the configuration is either picked from the variable, or fetched from that function. If we inline the function here it is harder to understand what is going on.

As far as I see we have two options (feel free to add more if you see them):
1. Keep the function and perhaps point out that it is only intended for internal use (although in theory you _can_ use it from outside.
2. Inline the function, and live with the (IMO) slightly less readable option.

What option do you think is best? Or do you have another option?

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -267,19 +158,143 @@ function(add_child_image_from_source)
       )
   endforeach()
 
+  unset(${output_cmake_args} PARENT_SCOPE)
```
You don't need to `unset` first.
```suggestion
```

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -267,19 +158,143 @@ function(add_child_image_from_source)
 
+  unset(${output_cmake_args} PARENT_SCOPE)
+  set(${output_cmake_args} "${child_image_cmake_args}" PARENT_SCOPE)
+
```
```suggestion
```

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -293,7 +293,16 @@ function(add_child_image_from_source)
   message("\n=== child image ${ACI_NAME} - ${ACI_DOMAIN}${inherited} begin ===")
 
-  get_child_image_cmake_args(${ACI_NAME} image_cmake_args)
+  if (PRE_DEFINED_IMAGE_CMAKE_ARGS_${ACI_NAME})
```
Who is expected to set this variable and why ?
What problem does this setting solve.

And we have no description of this, if this is expected to be set by the user.

tejlmand replied:
Just noticed, this is being set right before `add_child_image()` is called here:
https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/894a3b1741129de36b88fcfc911373f740dc2110/cmake/multi_image.cmake#L294-L299

So to me, this looks more like a `add_child_image()` argument that has been hidden by using the CMake cache, making code harder than necessary to read.

Should this variable have been an argument to `add_child_image()` ?

hakonfam replied:
Good point, it looks much better as an optional argument.

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -293,7 +293,16 @@ function(add_child_image_from_source)
+  endif()
+
+  # Expose the CMake args for the current image in case we want to create a variant build
+  set(${ACI_NAME}_IMAGE_CMAKE_ARGS "${image_cmake_args}" CACHE INTERNAL "")
```
not sure I fully see the reason why we need to have a cache variable for this.

Does it need to be preserved between CMake runs ?
After all it's a setting we construct on each CMake execution.

I think we should set those as properties on the child image.
https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/894a3b1741129de36b88fcfc911373f740dc2110/cmake/multi_image.cmake#L477

For example like:
```
set_property(TARGET ${ACI_NAME}_subimage APPEND PROPERTY CMAKE_ARGS "${value}")
```

We can then keep a long list of `-D{key1}={val1};-D{key2}={val2};...` or a simple list of: `key1;key2;...;keyN` and then each key has a corresponding property as: `CMAKE_ARGS_{key}`.

That will keep things nicely separated and fixed to the target itself.
This further avoids polluting the CMake cache.

Note: we do similar with the shared properties, although we create a dedicated target for those instead of using the already defined target for the child image.
I don't recall why a dedicated custom target was chosen for sharing of vars, but doing so is also an option above, but I assume the reason might be because we do the actual sharing through a generated file which then defines the target.
https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/894a3b1741129de36b88fcfc911373f740dc2110/cmake/extensions.cmake#L326

hakonfam replied:
I agree, I will add a target property.

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
 
   unset(${output_cmake_args} PARENT_SCOPE)
   set(${output_cmake_args} "${child_image_cmake_args}" PARENT_SCOPE)
+endfunction()
```
I believe this line removal should have been done directly in this commit: https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/pull/1/commits/e5cd3d8bf5d74d24b0f0ece83026b44c3666aa14

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
+# Required arguments are:
+# SOURCE_DIR - The source dir of the base image.
+#
+# Optional argumnets are:
```
`arguments`
```suggestion
# Optional arguments:
```

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
+#
+# Optional argumnets are:
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+function(add_s1_variant)
```
`multi_image.cmake` is generally generic regarding child image, yet the `add_s1_variant()` is a very specific variant type.

Could we consider adjusting this ?

hakonfam replied:
As I see it, the only alternative is to place this in its own `s1.cmake` file. This file would then only work when included alongside `multi_image.cmake` (or at least in a CMakeLists.txt that includes both). Given that this is sub-optimal, I would argue that having the slightly "non-generic" `add_s1_variant` in `multi_image.cmake` is preferable.

Would you rather 1) create s1.cmake or 2) keep it in `multi_image.cmake` ? 

Note that you can in theory use this for building direct XIP images compatible with MCUboot as well (not only B0). Maybe we should change its name to `add_direct_xip_variant` ?

hakonfam replied:
Actually, I would suggest that we get rid of this function, and then we either pass the required config directly to `add_image_variant` or we can have an optional argument which lets you link the variant into S1.

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
+  endif()
+
+  set(s1_override_conf
+    ${ZEPHYR_NRF_MODULE_DIR}/subsys/bootloader/image/is_s1_image.conf)
```
this file only contains a single line of the form:
https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/894a3b1741129de36b88fcfc911373f740dc2110/subsys/bootloader/image/is_s1_image.conf#L1

Maybe we should simply pass on this knowledge to the build system and not be going through an extra config file.
Something looks wrong in having `IS_S1_IMAGE` as Kconfig.

I believe we should follow the same principle used with `IMAGE_NAME`, where `IMAGE_NAME` also works as an identifier that tell the current image that it is a child image.

https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/894a3b1741129de36b88fcfc911373f740dc2110/cmake/multi_image.cmake#L440
https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/894a3b1741129de36b88fcfc911373f740dc2110/cmake/multi_image.cmake#L7

hakonfam replied:
This makes much more sense. It also simplifies the process of appending the extra config since we no longer need to search-replace the OVERLAY_CONFIG.

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
+  # Pass the extra configuration fragment to variant image.
+  if (base_image_configuration MATCHES ".*OVERLAY_CONFIG")
+    string(REGEX REPLACE
+      "-DOVERLAY_CONFIG="
+      "-DOVERLAY_CONFIG=${AIV_EXTRA_CONFIG}\\\\;"
```
this looks fragile to me.

If we have a dedicated property target, then I believe we can avoid this regex and escaping cause we can be working directly on native CMake lists.

hakonfam replied:
With the changes done regarding https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/pull/1#discussion_r785955603 this is no longer needed.

tejlmand commented on subsys/bootloader/image/CMakeLists.txt:
```diff
@@ -4,6 +4,16 @@
 #
 
+if (CONFIG_IS_S1_IMAGE)
+  # When building the S1 variant of an image, the configuration of the s1 image
+  # is identical to the base image, except 'CONFIG_IS_S1_IMAGE'. This file is
+  # included if 'CONFIG_SECURE_BOOT' is set, which can be the case for the base
+  # image. However, the logic in this file should only be performed once, for
+  # the base image, not for the s1 variant. Hence, we have this check to avoid
+  # execution of this file for the s1 variant.
```
This might be true with current way s1 in included, but in theory I assume that if someone used `add_image_variant()` with `s1.conf` and additional `EXTRA_CONFIG` then this might not be true.

That said, I think for time being this is ok.

tejlmand commented on samples/CMakeLists.txt:
```diff
@@ -76,6 +76,11 @@ if (CONFIG_SECURE_BOOT)
         ${NRF_DIR}/subsys/bootloader/image/log_minimal.conf
         )
   endif()
+  if (CONFIG_BUILD_S1_VARIANT AND "${CONFIG_S1_VARIANT_IMAGE_NAME}" STREQUAL "app")
```
If the `S1_VARIANT_IMAGE_NAME` already has the name `app`, should we then be using `app_s1_image` instead of `s1_image` in the build folder ?

hakonfam replied:
`CONFIG_S1_VARIANT_IMAGE_NAME` is actually used to denote the name of what we now call the "base image". So if this `if` evaluates as true - it means that its the `app` image that we should build an S1 variant for, as opposed to an `mcuboot` image.

tejlmand replied:
ok, guess that should go for another discussion.

Using a string with a quasistatic value seems fragile to me:
https://github.com/SebastianBoe/fw-nrfconnect-nrf-1/blob/31ed5f32652953c651d362cebdb31479d11aaf7b/subsys/bootloader/Kconfig#L17-L20

but I think changing this to be outside the scope of this PR.

tejlmand replied:
no changes requested for now.

tejlmand commented on cmake/multi_image.cmake:
```diff
@@ -173,7 +173,130 @@ function(get_child_image_cmake_args child_image output_cmake_args)
+    # from the top level parent image (that is, 'app').
+
+    set(base_image_configuration "${${AIV_BASE_IMAGE}_IMAGE_CMAKE_ARGS}")
+  else()
```
if we use properties on the child image target and then create a pseudo target for the parent image with its settings, then we can treat the parent image like any other image and create a variant of it.

No need for having the variant code itself doing the differentiation.

tejlmand reviewed (changes requested):
I believe we should avoid polluting the CMake cache with variant information.

If preferred, I can provide a commit presenting my initial thoughts.
